### PR TITLE
Add `better-boolean-variable-names`

### LINKED ISSUE #2441: Rule proposal: `better-boolean-var-declaration`
### Description

There are many irregular variable declarations in some Boolean variable declarations.

For example

```js
const showModal = false
```

This value is obviously a Boolean value, but it cannot be seen from the variable.

Standard variable declaration can improve code readability

Boolean variables should start with 'is', 'was', 'has', 'can', 'should', etc.

### Fail

```js
const showModal = false

const greatThen = a > b

const completed = progress === 100

const setProperty = key in object

const stringType = typeof input === 'string'

// ...
```

### Pass

```js
const isShowModal = false

const isGreatThen = a > b

const isCompleted = progress === 100

const hasSetProperty = key in object

const isStringType = typeof input === 'string'
```

### Proposed rule name

better-boolean-var-declaration

### Additional Info

This rule has been implemented in my company project, it can't autofix, but it can provide suggestions. I think it is useful

If this is acceptable, I'd be happy to PR

<details>
<summary>See the test case</summary>

```js
const assert = require('node:assert')
const test = require('node:test')

const outdent = require('outdent')
const { RuleTester } = require('eslint')

const rule = require('./boolean-var')

const ruleTester = new RuleTester({
    languageOptions: {
        ecmaVersion: 'latest',
        parserOptions: {
            sourceType: 'module'
        }
    }
})

test('boolean-var', t => {
    assert.strictEqual(typeof rule.create, 'function')

    ruleTester.run('boolean-var', rule, {
        valid: [
            // Valid cases
            {
                code: 'var isCompleted = true'
            },
            {
                code: 'var _isCompleted = true'
            },
            {
                code: 'var __isCompleted = true'
            },
            {
                code: 'var isCompleted = Boolean(0)'
            },
            {
                code: 'var isCompleted = !0'
            },
            {
                code: 'var isCompleted = !!0'
            },
            {
                code: 'var willUpdate = true; var allowsUpdate = true',
                options: [{ prefixes: ['will', 'allows'] }]
            }
        ],

        invalid: [
            // Invalid cases
            {
                code: 'var completed = true',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'completed',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Completed\``,
                                output: `var ${prefix}Completed = true`
                            }
                        })
                    }
                ]
            },
            {
                code: 'var completed = "true"',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'completed',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Completed\``,
                                output: `var ${prefix}Completed = "true"`
                            }
                        })
                    }
                ]
            },
            {
                code: 'var SKIP_REMIND = true',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'SKIP_REMIND',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix.toUpperCase()}_SKIP_REMIND\``,
                                output: `var ${prefix.toUpperCase()}_SKIP_REMIND = true`
                            }
                        })
                    }
                ]
            },
            {
                code: outdent`
                  var completed = true

                  console.log(completed)

                  function foo (completed) {
                    console.log(completed)
                  }

                  function bar () {
                      console.log(completed)
                  }
                `,
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'completed',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Completed\``,
                                output: outdent`
                                  var ${prefix}Completed = true
              
                                  console.log(${prefix}Completed)
              
                                  function foo (completed) {
                                    console.log(completed)
                                  }
              
                                  function bar () {
                                      console.log(${prefix}Completed)
                                  }
                                `
                            }
                        })
                    }
                ]
            },
            {
                code: 'var completed = Boolean(0)',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'completed',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Completed\``,
                                output: `var ${prefix}Completed = Boolean(0)`
                            }
                        })
                    }
                ]
            },
            {
                code: 'var completed = !0',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'completed',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Completed\``,
                                output: `var ${prefix}Completed = !0`
                            }
                        })
                    }
                ]
            },
            {
                code: 'var completed = !!0',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'completed',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Completed\``,
                                output: `var ${prefix}Completed = !!0`
                            }
                        })
                    }
                ]
            },
            {
                code: 'class Man { detected = false }',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'detected',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Detected\``,
                                output: `class Man { ${prefix}Detected = false }`
                            }
                        })
                    }
                ]
            },
            {
                code: 'class Man { static detected = false }',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'detected',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Detected\``,
                                output: `class Man { static ${prefix}Detected = false }`
                            }
                        })
                    }
                ]
            },
            {
                code: 'class Man { #detected = false }',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'detected',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Detected\``,
                                output: `class Man { #${prefix}Detected = false }`
                            }
                        })
                    }
                ]
            },
            {
                code: 'const a = b && c',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'a',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}A\``,
                                output: `const ${prefix}A = b && c`
                            }
                        })
                    }
                ]
            },
            {
                code: 'const func = typeof fn === "function"',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'func',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Func\``,
                                output: `const ${prefix}Func = typeof fn === "function"`
                            }
                        })
                    }
                ]
            },
            {
                code: 'const defined = key in object',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'defined',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Defined\``,
                                output: `const ${prefix}Defined = key in object`
                            }
                        })
                    }
                ]
            },
            {
                code: outdent`
                  const defined = key in object

                  const obj = {
                    defined
                  }
                `,
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'defined',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Defined\``,
                                output: outdent`
                                  const ${prefix}Defined = key in object
              
                                  const obj = {
                                    defined: ${prefix}Defined
                                  }
                                `
                            }
                        })
                    }
                ]
            },
            {
                code: outdent`
                  const defined = key in object

                  export default { defined }
                `,
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'defined',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Defined\``,
                                output: outdent`
                                  const ${prefix}Defined = key in object
              
                                  export default { defined: ${prefix}Defined }
                                `
                            }
                        })
                    }
                ]
            },
            {
                code: outdent`
                  const isFoo = true

                  const defined = isFoo

                  export default { defined }
                `,
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'defined',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Defined\``,
                                output: outdent`
                                  const isFoo = true

                                  const ${prefix}Defined = isFoo
              
                                  export default { defined: ${prefix}Defined }
                                `
                            }
                        })
                    }
                ]
            }
        ]
    })
})

```
</details>

### PR DESCRIPTION
**Pull request by axetroy**
Close #2441

Perhaps `prefer-readable-boolean-variable-names` is a better name?

### CODE REVIEW

**Comment by axetroy** [comment]
@fregante Found the issue in the codebase, can I fix it directly in this PR, or in separate PR?

```
/home/runner/work/eslint-plugin-unicorn/eslint-plugin-unicorn/rules/expiring-todo-comments.js
Error:   181:9  error  Prefer readable Boolean variable names  unicorn/better-boolean-variable-names

/home/runner/work/eslint-plugin-unicorn/eslint-plugin-unicorn/rules/explicit-length-check.js
Error:   170:8  error  Prefer readable Boolean variable names  unicorn/better-boolean-variable-names

/home/runner/work/eslint-plugin-unicorn/eslint-plugin-unicorn/rules/filename-case.js
Error:   169:8  error  Prefer readable Boolean variable names  unicorn/better-boolean-variable-names

/home/runner/work/eslint-plugin-unicorn/eslint-plugin-unicorn/rules/no-for-loop.js
Error:   340:10  error  Prefer readable Boolean variable names  unicorn/better-boolean-variable-names
Error:   355:10  error  Prefer readable Boolean variable names  unicorn/better-boolean-variable-names

/home/runner/work/eslint-plugin-unicorn/eslint-plugin-unicorn/rules/no-keyword-prefix.js
Error:   60:8  error  Prefer readable Boolean variable names  unicorn/better-boolean-variable-names
Error:   71:8  error  Prefer readable Boolean variable names  unicorn/better-boolean-variable-names

/home/runner/work/eslint-plugin-unicorn/eslint-plugin-unicorn/rules/no-process-exit.js
Error:   15:8  error  Prefer readable Boolean variable names  unicorn/better-boolean-variable-names
Error:   24:6  error  Prefer readable Boolean variable names  unicorn/better-boolean-variable-names

/home/runner/work/eslint-plugin-unicorn/eslint-plugin-unicorn/rules/prefer-default-parameters.js
Error:   100:8  error  Prefer readable Boolean variable names  unicorn/better-boolean-variable-names

/home/runner/work/eslint-plugin-unicorn/eslint-plugin-unicorn/rules/prefer-spread.js
Error:   59:8  error  Prefer readable Boolean variable names  unicorn/better-boolean-variable-names

/home/runner/work/eslint-plugin-unicorn/eslint-plugin-unicorn/rules/prefer-ternary.js
Error:    40:8   error  Prefer readable Boolean variable names  unicorn/better-boolean-variable-names
Error:   137:10  error  Prefer readable Boolean variable names  unicorn/better-boolean-variable-names
Error:   217:9   error  Prefer readable Boolean variable names  unicorn/better-boolean-variable-names

/home/runner/work/eslint-plugin-unicorn/eslint-plugin-unicorn/test/utils/snapshot-rule-tester.mjs
Error:   144:9  error  Prefer readable Boolean variable names  unicorn/better-boolean-variable-names

✖ 15 problems (15 errors, 0 warnings)

Error: Process completed with exit code 1.
```

**Comment by fregante** [comment]
Note that I’m not a maintainer, I just help triage the issues and have lots of opinions. 😅 I’ll leave that to @fisker

**Comment by axetroy** [comment]
@fregante OK, can you give me some advice, should I check the function name?

eg.

```ts
function completed(): boolean {}

const isCompleted = completed();
```

After fix:

```ts
function isCompleted1(): boolean {}

const isCompleted = isCompleted1();
```

Applying this rule to functions seems to cause naming problems

**Comment by fregante** [comment]
Heh, good point. 

What should be called `isCompleted`? The function or the constant? Both make sense. 

In your (corner) case it’s only an issue because you’re “caching” the value. Generally Sindre has preferred appending an underscore when the exact name is unavailable, like `function_` and `arguments_`. So the constant would be `isCompleted_` or e.g. `isCompletedLocal`. I don’t have an exact solution to this. 

By the way I think it would `hasCompleted` or `isComplete`

File: docs/rules/better-boolean-variable-names.md
Comment (fregante) [inline docs/rules/better-boolean-variable-names.md:10]: `will` also.

File: docs/rules/better-boolean-variable-names.md
Comment (fregante) [inline docs/rules/better-boolean-variable-names.md:46]: Why? I think got is right. 

``` js
const gotMilk = true
```

File: docs/rules/better-boolean-variable-names.md
Comment (fregante) [inline docs/rules/better-boolean-variable-names.md:158]: This gives me doubts about this rule. Having to configure your own verbs is a smell to me. `allows` should be valid by default… maybe? Or should it be `doesAllowDeletion`? What situations are we missing?

I don’t really want a rule that bugs you for names that are fine and that might be difficult to get right.

File: rules/utils/rename-variable.js
Comment (sindresorhus) [inline rules/utils/rename-variable.js:102]: Don't prefix comment lines with `*`

File: docs/rules/better-boolean-variable-names.md
Comment (sindresorhus) [inline docs/rules/better-boolean-variable-names.md:1]:
```suggestion
# Prefer readable boolean variable names
```

File: docs/rules/better-boolean-variable-names.md
Comment (sindresorhus) [inline docs/rules/better-boolean-variable-names.md:10]:
```suggestion
Boolean variable names should start with `is`/`was`/`has`/`can`/`should` for clarity and readability.
```

File: docs/rules/better-boolean-variable-names.md
Comment (sindresorhus) [inline docs/rules/better-boolean-variable-names.md:155]: Tab-indentation.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: docs/rules/better-boolean-variable-names.md
Comment (axetroy) [reply docs/rules/better-boolean-variable-names.md]: How about we add more prefixes to covering most scenarios? eg.

```js
["is","was","has","had","will","would","can","could","should","shall","does","needs","must","includes","enables","supports"]
```

File: rules/utils/is-boolean.js
Comment (axetroy) [inline rules/utils/is-boolean.js:11]: I know there is a method `isBooleanNode`, but it is not comprehensive enough, and it cannot track the type of Identifier.

https://github.com/sindresorhus/eslint-plugin-unicorn/blob/461b01cd69c42396ee1bda7ad583c4e2739b5696/rules/utils/boolean.js#L32-L37

So, I need more advice, should I use the same method.

File: docs/rules/better-boolean-variable-names.md
Comment (axetroy) [reply docs/rules/better-boolean-variable-names.md:58]: **Here is the response from AI:**

The word "got" is generally not suitable as a prefix for Boolean variables because it does not conform to the conventions for naming Boolean variables. Boolean variable names typically use verbs or adverbs to clearly express their Boolean value, such as "is," "has," "can," etc.

Using "got" as a prefix may result in variable names that are not sufficiently clear, thereby reducing the readability and maintainability of the code. Therefore, it is recommended to avoid using "got" as a prefix for Boolean variables.

Better choices include the following prefixes:

- `is`: isUserLoggedIn
- `has`: hasAdminPrivileges
- `can`: canEditProfile
- `should`: shouldDisplayNotification
- `will`: willSendEmail
- `needs`: needsPasswordReset

These prefixes can more clearly express the meaning of Boolean variables.

**Comment by jwhitmarsh** [comment]
@axetroy is this PR abandoned? Or just moved elsewhere?

**Pull request closed without merging**
